### PR TITLE
NXP backend: added support for `aten.pad` with mode `reflect`

### PR DESCRIPTION
**Pull request by novak-vaclav**
### Summary

Added support for `aten.pad` with mode `reflect` using new Neutron MLIR flow.

### Test plan

tests can be manually run using `pytest -c /dev/null backends/nxp/tests/`

cc @robert-kalmar @JakeStevens @digantdesai @rascani @MartinPavella

### CODE REVIEW

**Comment by novak-vaclav** [comment]
Fixed minor issues I found out about after creating the PR.

**Comment by novak-vaclav** [comment]
And rebased onto current main to run the tests on Neutron Software 3.2.0

File: backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        return padding_tfl
+
+    def convert(self, node: Node):
+        """Convert `aten.pad.default` to a TFLite padding operator."""
+        self.assert_convertible(node)
+
+        t_op = self._create_tflite_op_with_io_tensors(node)
+
+        x = t_op.tmp_inputs[0]
+        y = t_op.tmp_outputs[0]
```
Comment (roman-janik-nxp) [inline backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py]: Useless code.

File: backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+        paddings = self._convert_paddings_to_tflite(node.args[1], x)
+
+        paddings_tensor = self.builder.create_tensor_for_data(
+            np.asarray(paddings, "int32"), "paddings"
+        )
+
+        t_op.builtin_options = mirror_pad_options.MirrorPad(MirrorPadMode.REFLECT)
+        t_op.tmp_inputs = [x, paddings_tensor]
+        t_op.tmp_outputs = [y]
```
Comment (roman-janik-nxp) [inline backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py]: Useless code.

File: backends/nxp/neutron_partitioner.py [hunk 3]
```diff
@@ -538,6 +539,15 @@ def check_op_support_extended(node: torch.fx.Node):
                 )
                 return False
 
+            # under certain conditions an op cannot be decomposed further because there is
+            # no way to delegate the simpler ops
+            decomposable = node_converter.can_be_decomposed(
+                node, self.neutron_target_spec, self.custom_delegation_options
+            )
+            if not decomposable:
+                return True
```
Comment (roman-janik-nxp) [inline backends/nxp/neutron_partitioner.py]: I think it should be:
```
if decomposable:
                return False
```
This way if `decomposable == True` the `self.check_op_support(node)` can happen and if `decomposable == False` the `node_converter.is_supported()` can happen.

Or maybe the whole `can_be_decomposed` logic just move to `_is_supported_on_target()`.

File: backends/nxp/tests/ir/converter/node_converter/test_constant_pad_nd_converter.py [hunk 4]
```diff
@@ -79,15 +76,15 @@ def test__basic_nsys_inference(
         self, mocker, request, input_shape, paddings, use_qat
     ):
         # These test cases are also supported by the old flow.
```
Comment (roman-janik-nxp) [inline backends/nxp/tests/ir/converter/node_converter/test_constant_pad_nd_converter.py]: Obsolete comment I think.

File: backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            supported_types=[torch.int8, torch.uint8],
+            input_indices=[0],
+            output_indices=[0],
+        ):
+            return False
+
+        return True
+
+    @staticmethod
+    def _convert_paddings_to_tflite(
```
Comment (roman-janik-nxp) [inline backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py:67]: Why not use the `torch_padding_to_tflite_explicit_padding()` from `aten_translator.py`?

File: backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        # `constant` mode is decomposed to `aten.constant_pad_nd`.
+        # Conversion of `replicate`/`circular` Torch padding to TFLite
+        # is more complicated and is skipped for now.
+        if mode != "reflect":
+            return False
+
+        paddings = node.args[1]
+
+        # https://github.com/pytorch/pytorch/blob/v2.4.0/aten/src/ATen/native/PadNd.cpp#L38-L40
+        if len(paddings) > (input_rank(node, 0) * 2):
```
Comment (roman-janik-nxp) [inline backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py]: This and the following condition fails in torch already. I don't think we should include check for torch model validation.

File: backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py (hunk 5 above)
Comment (novak-vaclav) [reply backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py:67]: This cannot be done since I would need to pass TFLite input tensor to `torch_padding_to_tflite_explicit_padding` to permute the padding based on format:
```
if input_tensor.tensor_format.is_channels_last():
            # Permute padding to match tensor format.
            to_tflite_perm = create_channels_first_to_channels_last_permutation(
                input_tensor.rank
            )
            padding_tfl = apply_permutation_to(padding_tfl, to_tflite_perm)
```
But `convert_padding`, internally calling `torch_padding_to_tflite_explicit_padding`, is used in `_is_supported_in_IR`, where TFLite tensors don't exist yet.
`torch_padding_to_tflite_explicit_padding` could be refactored to account for this, but it would make the code of the function too cluttered imo.

File: backends/nxp/neutron_partitioner.py (hunk 3 above)
Comment (novak-vaclav) [reply backends/nxp/neutron_partitioner.py]: I moved the `can_be_decomposed` logic into `_is_supported_in_IR`, that makes more sense. Thank you.
I also took the liberty to refactor the whole `ops_not_to_decompose` method. The name and signature has to be kept the same since it is an overload, however I put some of the functionality into a separate private method.

**Comment by novak-vaclav** [comment]
Resolved comments from code review, resolved conflicts with main and rebased.

File: backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+from executorch.backends.nxp.backend.ir.tflite_generator.builtin_options import (
+    mirror_pad_options,
+)
+from executorch.backends.nxp.backend.neutron_target_spec import NeutronTargetSpec
+from torch.fx import Node
+from torch.nn import Parameter
+
+
+class PadConverter(NodeConverter):
+    """Convert `aten.pad.default` to a TFLite padding operator.
```
Comment (roman-janik-nxp) [inline backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py]: Please move this comment to `convert()` function as it is common for all other ops.

**Review by roman-janik-nxp — APPROVED** [review]
Just move the comment, otherwise GJ!

File: backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py (hunk 7 above)
Comment (novak-vaclav) [reply backends/nxp/backend/ir/converter/node_converters/ops_converters/pad_converter.py]: Fixed

**Comment by novak-vaclav** [comment]
Fixed all found issues in code review, rebased onto `main` and when tests pass, I will merge.

**Comment by novak-vaclav** [comment]
Github CI seemed to fail for some unrelated reason, so I rebased onto main and will wait till tests pass.

**Comment by novak-vaclav** [comment]
Internal tests are [passing](https://bamboo3.sw.nxp.com/browse/MLTECE-EXGHSPEC20-5) and Github CI failures do not seem to be related to this code. Merging.
